filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,285 @@
+		outlet:        out,
+		consumerGroup: consumerGroup,
+		sessionState:  sessionState,
+		log:           logp.NewLogger("kafka input").With("hosts", config.Hosts),
+	}
+
+	return input, nil
+}
+
+// A helper to safely use the current sarama session for the duration of the
+// given callback. Used when ACKing messages outside the body of the main
+// sarama callbacks. The session parameter may be nil if there is no active
+// session.
+func (state *kafkaSessionState) accessSession(
+	fn func(session sarama.ConsumerGroupSession),
+) {
+	state.mutex.Lock()
+	state.waitGroup.Add(1)
+	session := state.session
+	state.mutex.Unlock()
+	defer state.waitGroup.Done()
+	fn(session)
+}
+
+// A helper to safely set the session field after waiting on any pending
+// operations.
+func (state *kafkaSessionState) setSession(sess sarama.ConsumerGroupSession) {
+	state.mutex.Lock()
+	// Once we claim the mutex we still wait for any pending ACKs to be
+	// sent. (These may well fail if the session is ending, but that's better
+	// than calling a stale pointer.)
+	state.waitGroup.Wait()
+	state.session = sess
+	state.mutex.Unlock()
+}
+
+// Run starts the input by scanning for incoming messages and errors.
+func (input *kafkaInput) Run() {
+	input.runOnce.Do(func() {
+		// Track errors
+		go func() {
+			for err := range input.consumerGroup.Errors() {
+				input.log.Errorw("Error reading from kafka", "error", err)
+			}
+		}()
+
+		go func() {
+			for {
+				handler := groupHandler{input: input}
+
+				// Sarama uses standard go contexts to control cancellation, so we need
+				// to wrap our input context channel in that interface.
+				err := input.consumerGroup.Consume(
+					doneChannelContext(input.context.Done), input.config.Topics, handler)
+				if err != nil {
+					input.log.Errorw("Kafka consume error", "error", err)
+				}
+
+				// If Consume returned because the context was cancelled, don't resume.
+				select {
+				case <-input.context.Done:
+					return
+				default:
+				}
+			}
+		}()
+	})
+}
+
+// Wait shuts down the Input by cancelling the internal context.
+func (input *kafkaInput) Wait() {
+	input.Stop()
+	// Wait for the consumer group to shut down
+	input.consumerGroup.Close()
+}
+
+// Stop shuts down the Input by cancelling the internal context.
+func (input *kafkaInput) Stop() {
+	close(input.context.Done)
+}
+
+func arrayForKafkaHeaders(headers []*sarama.RecordHeader) []interface{} {
+	array := []interface{}{}
+	for _, header := range headers {
+		array = append(array, common.MapStr{
+			"key":   header.Key,
+			"value": header.Value,
+		})
+	}
+	return array
+}
+
+// A barebones implementation of context.Context wrapped around the done
+// channels that are more common in the beats codebase. This could be added
+// as a utility in a shared part of the code, but right now it's a special
+// case for sarama which requires a context.Context, so it's private until
+// there's at least one other use case.
+type channelCtx <-chan struct{}
```

Comment:
Please create a follow up issue to clear this in the future. I think the S3 input is adding the same type.

@ph did add some close signaling pattern to go-concert: https://github.com/elastic/go-concert/pull/7
It's more focused on signaling shutdown by building a tree of dependencies + calling functions during shutdown if required (e.g. to close sockets). I think it would make sense for his type to be castable to context.Context.
Consider to add this pattern for reuse to go-concert. Historically we have a many done channels all over the place. Would be nice if we can pass them down into IO libs as context.Context from other places as well.